### PR TITLE
Integration of YOLOv8-face close #732

### LINKED ISSUE #732: I added support for YOLO-v8 Faces as a face detector
Hello, first of all i apologize because i'm new to coding and this code is ugly, but it does what i need it to do. 
I also apologize for not using the pull request function, but i don't think my code meets the quality criteria for that, I also don't know how to use it.
I'm working on a project for video analysis, and i need to encode many frames, the more frames i can encode, the better it is (the frames are already extracted so all i need is something as fast as possible to go over each frame and extract the faces encodings).
I started with face_recognition and managed to subsitute its native detector with yolo, increasing its speed by more than a order of magnitude. Then I decided to try to use deepface hoping to find it faster, but I didn't. So i decided to try and implement it also here, and... well it works and it's blazing fast, in my experience, faster than any other detector in your suite.
Sadly though, it's slower than my face_recognition implementation for all the encoding models except dlib, which is slight faster. For some reason, DBSCAN also gives me worse results with every model, so i'm not sure i can move away from face_recognition.

Anyway, I managed to implement yolo-v8 face (from here: https://github.com/derronqi/yolov8-face) into your library. To do this, as i already said, i just took dlib's face detector's file (XD) and overwrote it with this code i wrote:

```
import os
import gdown
from deepface.detectors import FaceDetector
from deepface.commons import functions

def build_model():
  home = functions.get_deepface_home()
  if os.path.isfile(home + "/.deepface/weights/yolov8n-face.pt") != True:
    url = "https://drive.google.com/uc?id=1qcr9DbgsX3ryrz2uU8w4Xm3cOrRywXqb"
    output = home + "/.deepface/weights/yolov8n-face.pt"
    gdown.download(url, output, quiet=False)
    print("Downloaded YOLO model yolo8vn-face.pt")
  from ultralytics import YOLO
  face_detector = YOLO(home + "/.deepface/weights/yolov8n-face.pt")
  print("YOLO ready")
  return face_detector


def detect_face(face_detector, img, align=False):

  resp = []
  align=False
  confidence = 0.95
  detected_face = None
  detections = [tuple(int(x) for x in sublist) for sublist in face_detector(img, verbose=False)[0].boxes.xywh.tolist()]

  if len(detections) > 0:

      for detection in detections:
          x,y,w,h = detection[0], detection[1], detection[2], detection[3]
          img_region = [x, y, w, h]
          detected_face = img[int(y) : int(y + h), int(x) : int(x + w)]
          resp.append((detected_face, img_region, confidence))

  return resp
```
I know it's a big hack with the confidence and align, but I didn't know what those do and just decided to try and ignore them since I don't need them for my project.
I hope I can be of help or maybe you can improve the code or even give me directions to achieve faster speeds. By the way your new video about the facebook model is great, i'm going to try and replicate it in my project.

### PR DESCRIPTION
**Pull request by Vincent-Stragier**
Hello,

Here is an implementation of the integration of YOLOv8 in DeepFace (related to #732).

This uses the model weights provided by YOLOv8 face ([here](https://github.com/derronqi/yolov8-face)).
The `ultralytics` python library is needed to run the model. Only one model is working (yolov8n), see https://github.com/derronqi/yolov8-face/issues/3, for more information.

The confidence is extracted, as well as landmarks (the confidences of the landmarks is also used to decide to realize or not the alignment) to align the face.

Regards,
Vincent

### CODE REVIEW

**Comment by serengil** [comment]
Make some comments

**Review by serengil — COMMENTED** [review]
leaving some comments

File: .vscode/settings.json [hunk 1]
```diff
@@ -9,5 +9,7 @@
   "python.formatting.provider": "black",
   "python.formatting.blackArgs": ["--line-length=100"],
   "editor.fontWeight": "normal",
-  "python.analysis.extraPaths": ["./deepface"]
+  "python.analysis.extraPaths": ["./deepface"],
```
[the remark this answers is not in the document]
Comment (Vincent-Stragier) [reply .vscode/settings.json]: Changes reverted.

File: deepface/DeepFace.py [hunk 2]
```diff
@@ -119,9 +117,9 @@ def verify(
 
             detector_backend (string): set face detector backend to opencv, retinaface, mtcnn, ssd,
             dlib or mediapipe
-            
+
```
[the remark this answers is not in the document]
Comment (Vincent-Stragier) [reply deepface/DeepFace.py:120]: I removed the unwanted line return, but I think the other modification in the docstrings should stay (`dlib or mediapipe` --> `dlib, mediapipe or yolov8n.`).

File: deepface/detectors/OpenCvWrapper.py [hunk 3]
```diff
@@ -45,26 +45,29 @@ def detect_face(detector, img, align=True):
     img_region = [0, 0, img.shape[1], img.shape[0]]
 
     faces = []
+    scores = []
```
[the remark this answers is not in the document]
Comment (Vincent-Stragier) [reply deepface/detectors/OpenCvWrapper.py:49]: Ok, I'll revert that file and do another PR.

File: deepface/detectors/Yolov8faceWrapper.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+from deepface.detectors import FaceDetector
+
+PATHS = {
+    "yolov8-lite-t": "/.deepface/weights/yolov8-lite-t.pt",
+    "yolov8-lite-s": "/.deepface/weights/yolov8-lite-s.pt",
+    "yolov8n": "/.deepface/weights/yolov8n-face.pt",
```
[the remark this answers is not in the document]
Comment (Vincent-Stragier) [reply deepface/detectors/Yolov8faceWrapper.py:7]: Well, all the models exit, however, only `yolov8n` works for some reason. I'll comment them out and add a comment.

[the remark this answers is not in the document]
Comment (serengil) [reply deepface/detectors/Yolov8faceWrapper.py:7]: Then please remove broken ones. We should just have the working one.

[the remark this answers is not in the document]
Comment (Vincent-Stragier) [reply deepface/detectors/Yolov8faceWrapper.py:5]: OK, they've been removed.

**Comment by Vincent-Stragier** [comment]
@serengil it seems mainly resolved to me (nearly ready to merge). I still have some doubt about the CamelCasing for the module. I think it could be `YoloV8FaceWrapper` or `YOLOV8FaceWrapper` instead of `Yolov8faceWrapper`.

File: deepface/DeepFace.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (serengil) [reply deepface/DeepFace.py:120]: sure, this should be available in the docstring

File: deepface/detectors/Yolov8faceWrapper.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+PATHS = {
+    "yolov8n": "/.deepface/weights/yolov8n-face.pt",
+}
+
+# Google Drive base URL
+BASE_URL = "https://drive.google.com/uc?id="
+
+# Models' Google Drive IDs
+IDS = {
+    "yolov8n": "1qcr9DbgsX3ryrz2uU8w4Xm3cOrRywXqb",
```
Comment (serengil) [inline deepface/detectors/Yolov8faceWrapper.py:13]: please do not use IDS here. you can directly use this is in weights.

File: README.md [hunk 6]
```diff
@@ -207,7 +207,8 @@ backends = [
   'dlib', 
   'mtcnn', 
   'retinaface', 
-  'mediapipe'
+  'mediapipe',
+  'yolov8n',
```
Comment (serengil) [inline README.md:211]: IMO, detector name should be yolo instead of yolov8n

File: deepface/detectors/FaceDetector.py [hunk 7]
```diff
@@ -23,6 +23,7 @@ def build_model(detector_backend):
         "mtcnn": MtcnnWrapper.build_model,
         "retinaface": RetinaFaceWrapper.build_model,
         "mediapipe": MediapipeWrapper.build_model,
+        "yolov8n": Yolov8faceWrapper.build_model("yolov8n"),
```
Comment (serengil) [inline deepface/detectors/FaceDetector.py:26]: please just call build model and do not pass "yolov8n" argument. that detector will build default model.

File: deepface/detectors/FaceDetector.py [hunk 8]
```diff
@@ -9,11 +9,11 @@
     MtcnnWrapper,
     RetinaFaceWrapper,
     MediapipeWrapper,
+    Yolov8faceWrapper,
```
Comment (serengil) [inline deepface/detectors/FaceDetector.py:12]: IMO, name should be YoloWrapper

File: deepface/detectors/Yolov8faceWrapper.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    func_weights_path = f"{get_deepface_home()}{PATHS[model]}"
+    func_url = f"{BASE_URL}{IDS[model]}"
+
+    # Define function to build the model
+    def _build_model(weights_path: str = func_weights_path, url: str = func_url):
+        import gdown
+        import os
+
+        # Import the Ultralytics YOLO model
+        from ultralytics import YOLO
```
Comment (serengil) [inline deepface/detectors/Yolov8faceWrapper.py:31]: this import requires to download a ultralytics python package. this should be mentioned in requirements.txt or requirements_additional.txt. if this is not a must dependency, then add it to requirements_additional.

File: deepface/detectors/Yolov8faceWrapper.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        confidence = result.boxes.conf.tolist()[0]
+
+        x, y, w, h = int(x - w / 2), int(y - h / 2), int(w), int(h)
+        detected_face = img[y: y + h, x: x + w].copy()
+
+        if align:
+            # Extract landmarks
+            left_eye, right_eye, _, _, _ = result.keypoints.tolist()
+            # Check the landmarks confidence before alignment
+            if left_eye[2] > 0.5 and right_eye[2] > 0.5:
```
Comment (serengil) [inline deepface/detectors/Yolov8faceWrapper.py:64]: confidence score 0.5 should be defined with a variable in the module level.

File: README.md (hunk 6 above)
Comment (Vincent-Stragier) [reply README.md:211]: YOLO evolve in all direction, so I'm not sure if it's ideal to not specify the version

File: deepface/detectors/Yolov8faceWrapper.py (hunk 5 above)
Comment (Vincent-Stragier) [reply deepface/detectors/Yolov8faceWrapper.py:13]: ok

File: deepface/detectors/FaceDetector.py (hunk 7 above)
Comment (Vincent-Stragier) [reply deepface/detectors/FaceDetector.py:26]: ok

File: deepface/detectors/Yolov8faceWrapper.py (hunk 9 above)
Comment (Vincent-Stragier) [reply deepface/detectors/Yolov8faceWrapper.py:31]: Probably not a must (like mediapipe). I have to add a meaningful import error.

File: deepface/detectors/FaceDetector.py (hunk 8 above)
Comment (Vincent-Stragier) [reply deepface/detectors/FaceDetector.py:12]: ok

File: deepface/detectors/Yolov8faceWrapper.py (hunk 10 above)
Comment (Vincent-Stragier) [reply deepface/detectors/Yolov8faceWrapper.py:64]: Like in the YoloWrapper?

```python
LANDMARKS_CONFIDENCE_THRESHOLD = 0.5
```

Comment (serengil) [reply deepface/detectors/Yolov8faceWrapper.py]: yes this seems good

Comment (Vincent-Stragier) [reply deepface/detectors/Yolov8faceWrapper.py]: ok

File: README.md (hunk 6 above)
Comment (serengil) [reply README.md:211]: yolov8 instead of yolov8n?

Comment (Vincent-Stragier) [reply README.md]: Ok with `yolov8`

**Comment by serengil** [comment]
LGTM

**Comment by serengil** [comment]
a ticket raised about yolo - https://github.com/serengil/deepface/issues/788

@Vincent-Stragier is this meaningful for you?

**Pull request merged**